[TF:TRT] Enable dynamic batch dim for Conv2dBackpropInput

christopherbate opened the pull request:
This PR enables dynamic batch dimension for Conv2dBackpropInput.

We enable dynamic batch dimension for Conv2DBackpropInput simply by relaxing the check to allow dynamic batch dimension in the converter. We also add positive and negative tests.

The support for dynamic height and width will come later. This is because we will need to add an additional dynamic padding operation in order to support dynamic height and width for Conv2dBackpropinput.

christopherbate commented:
@bixia1 
@tfeher

tfeher commented:
Thanks @christopherbate I had a quick look at the changes, it looks good overall.

bixia1 commented:
This is what you have for the PR description:
This PR enables dynamic batch dimension for Conv2dBackpropInput.

Full dynamic shape for height and width will come later. We first need to add an additional dynamic padding operation support in order to handle the case where the input_shapes operation argument is non-constant.

Here is my suggestion to revise it. Do you see any problem?
This PR enables dynamic batch dimension for Conv2dBackpropInput.

We enable dynamic batch dimension for Conv2DBackpropInput simply by relaxing the check to allow dynamic batch dimension in the converter. We also add positive and negative tests.

The support for dynamic height and width will come later. This is because we will need to add an additional dynamic padding operation in order to support dynamic height and width for Conv2dBackpropinput.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2123,11 +2123,18 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
-    if (!HasStaticShape(tensor->getDimensions())) {
-      // TODO(tfeher): Allow dynamic input. We need to implement padding
-      // correction for dynamic shapes in this case.
+
+    auto has_dynamic_hw_shape = [](const nvinfer1::Dims& dims) {
+      int start_idx = dims.nbDims == 4 ? 2 : 1;
+      for (int i = start_idx; i < dims.nbDims; ++i) {
+        if (dims.d[i] < 0) return true;
+      }
+      return false;
+    };
+    if (has_dynamic_hw_shape(tensor->getDimensions())) {
```
(1) Can you add comment to explain the situations for dims.nbDims=4 and dims.nbDims!=4? I was understand the impression that nbDims is always 4 here.  For nbDims!=4, what is the data format? NHW?
(2) I don't see a test case for nbDims !=4. Am I right? Please add a test case for that.
(3) The lambda here doesn't look necessary as it is only used once.

christopherbate replied:
In the code it's very confusing, in some situations we use TRT style implicit batch dims, other times use a batch dimension is inserted, etc. I'll go back and check.

christopherbate replied:
The converter receives the input `tensor` through the input params. `tensor` is a `TensorProxyPtr`. It either has 3 dims (implicit batch) or 4 dims (explicit batch). You can see further down lines in some pre-existing lines the following code:

```
const auto tensor_dim = tensor->getDimensions();
const int c_dim_size = tensor_dim.d[params->use_implicit_batch ? 0 : 1];
```

My understanding is that in the 2D operations case, `use_implicit_batch` and `tensor_dim.nbDims != 4` are the same. `tensor_dims.nbDims==4` means `!use_implicit_batch`

bixia1 replied:
Can we write the code like this:
  if (params->use_implicit_batch) {
     return error if dims.nbDims != 3
     start_idx = 1
 } else {
    return error if dims.nbDims != 4
    start_idx = 2
 }

christopherbate replied:
Yes of course, I'll change it

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -5052,28 +5054,81 @@ TEST_P(OpConverter_FP32_Test, ConvertConv2DBackpropInput) {
-        1});
+        /*concrete input dims=*/{1, 1, 2, 2}});
+
+    // These params test dynamic size with respect to batch dimension.
```
Let's revise the comment:
Test dynamic batch dimension.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -5052,28 +5054,81 @@ TEST_P(OpConverter_FP32_Test, ConvertConv2DBackpropInput) {
+                   /*conversion_status=*/Status::OK(),
+                   /*concrete input dims=*/{2, 1, 2, 2}});
+
+    // Dynamic size with respect to the height and width of input is not yet
+    // supported due to differences in Tensorflow and TensorRT.
```
Let's revise this comment to:
Test dynamic height and width.

It is not necessary to say whether the case is supported or not. In the future, when this is supported, we will just change the expected results without having to modify this comment.

christopherbate commented:
> This is what you have for the PR description:
> This PR enables dynamic batch dimension for Conv2dBackpropInput.
> 
> Full dynamic shape for height and width will come later. We first need to add an additional dynamic padding operation support in order to handle the case where the input_shapes operation argument is non-constant.
> 
> Here is my suggestion to revise it. Do you see any problem?
> This PR enables dynamic batch dimension for Conv2dBackpropInput.
> 
> We enable dynamic batch dimension for Conv2DBackpropInput simply by relaxing the check to allow dynamic batch dimension in the converter. We also add positive and negative tests.
> 
> The support for dynamic height and width will come later. This is because we will need to add an additional dynamic padding operation in order to support dynamic height and width for Conv2dBackpropinput.

No problem,  I updated the description.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2120,11 +2120,34 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
+    bool has_dynamic_hw_shape{false};
+    {
+      int start_idx{0};
+      auto dims = tensor->getDimensions();
```
I don't see how dims is used.

christopherbate replied:
It's used in two statements `if(dims.nbDims != ....)` which you recommended.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2120,11 +2120,34 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
-      // correction for dynamic shapes in this case.
+
+    bool has_dynamic_hw_shape{false};
+    {
```
Can we avoid using this nested scope?

christopherbate replied:
fixed

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2120,11 +2120,34 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
+          return errors::Internal(
+              "In explicit batch mode, input nbDims should be 4");
+        }
+        start_idx = 0;
```
do you actually want "start_idx = 2" instead?

christopherbate replied:
Sorry I fat fingered a couple things. I just pushed an update

christopherbate committed:
TF-TRT: Enable dynamic batch dim for Conv2dBackpropInput

christopherbate commented:
I updated test parameter naming to be inline with other dynamic shape developments.

I've rebased on master, but it still it shows difference between `third_party/tf_runtime/workspace.bzl`
